### PR TITLE
SOLR-11731: LatLonPointSpatialField could be improved to return the lat/lon from docValues

### PR DESCRIPTION
**Pull request by mrkarthik**

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/geo/GeoEncodingUtils.java [hunk 1]
```diff
@@ -127,6 +130,10 @@ public static double decodeLatitude(int encoded) {
   public static double decodeLatitude(byte[] src, int offset) {
     return decodeLatitude(NumericUtils.sortableBytesToInt(src, offset));
   }
+  
+  public static double decodeLatitudeCeil(long encoded) {
```
Comment (dsmiley) [inline lucene/core/src/java/org/apache/lucene/geo/GeoEncodingUtils.java]: @nknize  could you please review this part?  It's not clear to me why GeoEncodingUtils has dual ceil/floor for encode but not for decode.

Comment (dsmiley) [reply lucene/core/src/java/org/apache/lucene/geo/GeoEncodingUtils.java]: Or maybe you know @mrkarthik since you felt this addition here was needed?

Comment (mrkarthik) [reply lucene/core/src/java/org/apache/lucene/geo/GeoEncodingUtils.java]: @dsmiley i can add the other method to do the floor.

File: solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java [hunk 2]
```diff
@@ -71,6 +74,10 @@ protected SpatialStrategy newSpatialStrategy(String fieldName) {
     SchemaField schemaField = schema.getField(fieldName); // TODO change AbstractSpatialFieldType so we get schemaField?
     return new LatLonPointSpatialStrategy(ctx, fieldName, schemaField.indexed(), schemaField.hasDocValues());
   }
+  
+  public String geoValueToStringValue(long value) {
```
[the remark this answers is not in the document]
Comment (mrkarthik) [reply solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java]: will make the change and update the PR

File: solr/core/src/test/org/apache/solr/search/TestSolr4Spatial.java [hunk 3]
```diff
@@ -157,7 +158,7 @@ public void testIntersectFilter() throws Exception {
   @Test
   public void checkResultFormat() throws Exception {
     //Check input and output format is the same
-    String IN = "89.9,-130";//lat,lon
```
[the remark this answers is not in the document]
Comment (mrkarthik) [reply solr/core/src/test/org/apache/solr/search/TestSolr4Spatial.java]: No not required, i will remove the change.

File: solr/core/src/test/org/apache/solr/search/TestSolr4Spatial2.java [hunk 4]
```diff
@@ -122,19 +122,19 @@ public void testRptWithGeometryGeo3dField() throws Exception {
...
     assertU(adoc("id", "0",
         "llp_1_dv_st", "-75,41",
-        "llp_1_dv", "-80,20",
-        "llp_1_dv_dvasst", "10,-30"));
+        "llp_1_dv", "-80.0,20.0",
+        "llp_1_dv_dvasst", "40.299599,-74.082728"));
     assertU(commit());
     assertJQ(req("q","*:*", "fl","*"),
         "response/docs/[0]/llp_1_dv_st=='-75,41'",
         // Right now we do not support decoding point value from dv field
```
[the remark this answers is not in the document]
Comment (mrkarthik) [reply solr/core/src/test/org/apache/solr/search/TestSolr4Spatial2.java]: will change the comment, llp_1_dv is not retrieved stored="false" and useDocValuesAsStored="false"

**Comment by mrkarthik** [comment]
@dsmiley Updated the PR based on the comments.

File: solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java [hunk 5]
```diff
@@ -486,16 +486,14 @@ private Object decodeDVField(int localId, LeafReader leafReader, String fieldNam
...
-            long number = numericDv.nextValue();
-            return ((LatLonPointSpatialField) schemaField.getType()).geoValueToStringValue(number);
-          }
           final List<Object> outValues = new ArrayList<>(numericDv.docValueCount());
           for (int i = 0; i < numericDv.docValueCount(); i++) {
             long number = numericDv.nextValue();
             Object value = decodeNumberFromDV(schemaField, number, true);
             // return immediately if the number is not decodable, hence won't return an empty list.
             if (value == null) return null;
+            // return the value as "lat, lon" if its not multi-valued
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java]: This is not consistent with how Solr normally does things.  If the field type is declared as multiValued then we normally always return as a list; otherwise we never do.  Here I think you're making that vary per document dependent on how many values the document has.

File: solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java [hunk 6]
```diff
@@ -75,8 +77,16 @@ protected SpatialStrategy newSpatialStrategy(String fieldName) {
     return new LatLonPointSpatialStrategy(ctx, fieldName, schemaField.indexed(), schemaField.hasDocValues());
   }
   
-  public String geoValueToStringValue(long value) {
-    return new String(decodeLatitudeCeil(value) + "," + decodeLongitudeCeil(value));
+  /**
+   * Converts to "lat, lon"
+   * @param value Non-null; stored location field data
+   * @return Non-null; "lat, lon" with 6 decimal point precision
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java]: Why 6 decimal points?  Is that sufficient to represent the data to as much precision as is decoded?  Perhaps instead of putting the constant '6' in the code, it should be calculated so that we can see how 6 is arrived at.  What does that translate to in the metric system?

File: solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java [hunk 7]
```diff
@@ -75,8 +77,16 @@ protected SpatialStrategy newSpatialStrategy(String fieldName) {
...
+   */
+  public static String decodeDocValueToString(long value) {
+    double latitudeDecoded = BigDecimal.valueOf(decodeLatitude((int) (value >> 32))).setScale(6, HALF_UP).doubleValue();
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java]: Lets have some comments explaining why this algorithm here is what it is.  Why HALF_UP?

Can we skip the doubleValue and just do toPlainString (avoiding exponent notation of toString) since we're composing a string in the end?  In other words, lets avoid the pointless double primitive intermediary.

File: solr/core/src/test/org/apache/solr/search/TestSolr4Spatial2.java [hunk 8]
```diff
@@ -120,21 +120,27 @@ public void testRptWithGeometryGeo3dField() throws Exception {
   
   @Test
   public void testLatLonRetrieval() throws Exception {
-    assertU(adoc("id", "0",
-        "llp_1_dv_st", "-75,41",
-        "llp_1_dv", "-80.0,20.0",
-        "llp_1_dv_dvasst", "40.299599,-74.082728"));
+    assertU(adoc("id", "0", "llp_1_dv_st", "-75,41")); // stored
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/search/TestSolr4Spatial2.java]: Please also test -90 and +90, -180 and +180.  This will help ensure there aren't edge cases (literally) in the DV decoding logic.

File: solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java (hunk 5 above)
Comment (mrkarthik) [reply solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java]: The only reason I had to do this is LatLonDocValuesField type is SORTED_NUMERIC, so even for non-multivalued data, we would be returning an array.  If that is what is excepted?
If the field is stored then it would return string for non-multi-valued data and string array for multi-valued data.

File: solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java (hunk 6 above)
Comment (mrkarthik) [reply solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java]: 6 was just based on what i read, For 40.299599,-74.082728 it is 40°17'58.52",74°4'57.79".  I will remove it

File: solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java (hunk 7 above)
Comment (mrkarthik) [reply solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java]: HALF_UP is only for ceil, I will remove the rounding.

File: solr/core/src/test/org/apache/solr/search/TestSolr4Spatial2.java (hunk 8 above)
Comment (mrkarthik) [reply solr/core/src/test/org/apache/solr/search/TestSolr4Spatial2.java]: Sure. will add

**Comment by dsmiley** [comment]
In general my review here has a higher attention to detail than normal because I think there are some nitty gritty details here that should be explained so that not only me but future reviewers know _why_ we're doing what we're doing.  Spatial / numerical stuff demands more attention to detail.

File: solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java (hunk 7 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java]: I'm not arguing with your choice; just trying to understand.

File: solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java (hunk 6 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java]: By "What does that translate to in the metric system?" I mean for example how many meters would this translate to with 6 decimal points?

"6 was just based on what i read"   read where?

File: solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java (hunk 5 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java]: I understand that internally the type is SORTED_NUMERIC either way.  But externally (from user's point of view) the visible behavior should be consistent with what would happen if the field were stored.  Please ensure this distinction is tested too (if it isn't already)

Comment (mrkarthik) [reply solr/core/src/java/org/apache/solr/search/SolrDocumentFetcher.java]: Yes, I have test case of both store and docValue

File: solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java (hunk 6 above)
Comment (mrkarthik) [reply solr/core/src/java/org/apache/solr/schema/LatLonPointSpatialField.java]: Based on the article it would be close 111mm
https://gis.stackexchange.com/a/8674
https://en.wikipedia.org/wiki/Decimal_degrees

**Comment by mrkarthik** [comment]
@dsmiley I have update the PR based on the feedback
